### PR TITLE
Edge

### PR DESCRIPTION
**Pull request by gsvarovsky**
- [x] Compliance tests run successfully
- [x] Documentation deployed at https://edge.js.m-ld.org/
- [x] Website deployed at https://edge.m-ld.org/ & smoke tested
- [x] No pre-release dependencies
- [x] m-ld/m-ld-nodejs-starter#edge smoke tested
- [x] m-ld/m-ld-web-starter#edge smoke tested
- [x] m-ld/m-ld-cli#edge smoke tested
- [x] Complete CHANGELOG.md

### CODE REVIEW

File: src/engine/dataset/index.ts [hunk 1]
```diff
@@ -21,6 +20,7 @@ import { MutableOperation } from '../ops';
 import { MeldError } from '../MeldError';
 import { BaseStream, CountableRdf, QueryableRdfSource } from '../../rdfjs-support';
 import { AllMetadataSupport, Query } from '@rdfjs/types';
+import { uuid } from '../../util';
```
Comment (Angus-McAllister) [inline src/engine/dataset/index.ts:21]: Couldn't help noticing during this import of `uuid` from `util`.  Given that this commit was intended to replace uuid with SHA1, I'm wondering whether this is actually correct?

File: src/engine/remotes/PubsubRemotes.ts [hunk 2]
```diff
@@ -13,7 +12,7 @@ import { Future, MsgPack, Stopwatch, toJSON } from '../util';
 import { delay, first, ignoreElements, map, reduce, tap, timeout, toArray } from 'rxjs/operators';
 import { MeldError, MeldErrorStatus } from '../MeldError';
 import { AbstractMeld } from '../AbstractMeld';
-import { MeldExtensions, MeldReadState, noTransportSecurity, shortId } from '../../index';
+import { MeldExtensions, MeldReadState, noTransportSecurity, shortId, uuid } from '../../index';
```
Comment (Angus-McAllister) [inline src/engine/remotes/PubsubRemotes.ts:16]: Noticed that `uuid` is imported from here too `util`.  As before, given that this commit was intended to replace uuid with SHA1, is this actually correct?

File: src/util.ts [hunk 3]
```diff
@@ -18,17 +17,14 @@ export function array<T>(value?: T | T[] | null): T[] {
...
+ * string with a fast hash.
  * @return a string identifier that is safe to use as an HTML (& XML) element Id
  * @category Utility
  */
-export function shortId(spec: number | string = 8) {
-  if (typeof spec == 'number') {
-    let d = new Date().getTime();
-    return ('a' + 'x'.repeat(spec - 1)).replace(/[ax]/g, c =>
-      ((d + Math.random() * 16) % (c == 'a' ? 6 : 16) + (c == 'a' ? 10 : 0) | 0).toString(16));
+export function shortId(spec?: string) {
```
Comment (Angus-McAllister) [inline src/util.ts:25]: Looks much more elegant!

**Comment by Angus-McAllister** [comment]
That's a lot of commits!  I've reviewed a couple at random, with comments, but overall I don't see any showstoppers with the merge.  May I suggest, however, that `master` be renamed `main`, for well-established reasons?

File: src/engine/dataset/index.ts (hunk 1 above)
Comment (gsvarovsky) [reply src/engine/dataset/index.ts:21]: This is correct. The switch from UUID is in the logical clock. Here it's being used for a guid.

File: src/engine/remotes/PubsubRemotes.ts (hunk 2 above)
Comment (gsvarovsky) [reply src/engine/remotes/PubsubRemotes.ts:16]: ditto

**Comment by gsvarovsky** [comment]
Please approve, thanks @Angus-McAllister

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,7 +1,37 @@
-Version with a major version number of 0 are developer pre-releases. Unless
-otherwise stated, pre-release versions are not backwards-compatible with prior
-versions for storage or network transmission, so they must be used with data
-persisted using the same version, and other clones of the same version.
+Version with a major version number of 0 are developer pre-releases.
```
Comment (Angus-McAllister) [inline CHANGELOG.md:1]: Should be version**s** (plural)

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,7 +1,37 @@
...
+
+- API: `ask` query in a read state to optimally discover if some pattern exists
+- API: `propertyValue` type options include explicit `Subject`, `Reference` and `VocabReference`
+  - **[breaking]**: `Object` is no longer allowed
+ 
+### [breaking]
+
+- Upgraded to the new generation of `*-level` packages (see [upgrade guide](https://github.com/Level/community#how-do-i-upgrade-to-abstract-level))
+  - `MeldMemDown` is no longer required or available, instead please use `MemoryLevel` from [`memory-level`](https://github.com/Level/memory-level).
+- The `dist` package, containing remotes and other extensions, has been renamed to `ext`. For example, now use `import '@m-ld/m-ld/ext/socket.io'`.
```
Comment (Angus-McAllister) [inline CHANGELOG.md:20]: That's a good idea; it makes it easier to understand the purpose of that package.

**Review by Angus-McAllister — APPROVED** [review]
Looks good - just one small grammatical error I spotted that needs attention, in CHANGELOG.md.

**Comment by gsvarovsky** [comment]
Holding off due to spec defect discovered in m-ld/m-ld-security-spec#4

**Pull request merged**